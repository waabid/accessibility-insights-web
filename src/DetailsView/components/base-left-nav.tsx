// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { TestRequirementLeftNavLink } from 'DetailsView/components/left-nav/assessment-left-nav';
import { INavLink, Nav } from 'office-ui-fabric-react';
import * as React from 'react';

export type onBaseLeftNavItemClick = (
    event: React.MouseEvent<HTMLElement>,
    item: BaseLeftNavLink,
) => void;
export type onTestRequirementClick = (
    event: React.MouseEvent<HTMLElement>,
    item: TestRequirementLeftNavLink,
) => void;
export type onBaseLeftNavItemRender = (
    link: BaseLeftNavLink,
    renderIcon: (link: BaseLeftNavLink) => JSX.Element,
) => JSX.Element;

export type BaseLeftNavProps = {
    selectedKey: string;
    links: BaseLeftNavLink[];
    renderIcon: (link: BaseLeftNavLink) => JSX.Element;
};

export interface BaseLeftNavLink extends INavLink {
    percentComplete?: number;
    onRenderNavLink: onBaseLeftNavItemRender;
    onClickNavLink: onBaseLeftNavItemClick;
}

export interface BaseLeftNavLinkProps {
    link: BaseLeftNavLink;
    renderIcon: (link: BaseLeftNavLink) => JSX.Element;
}

export class BaseLeftNav extends React.Component<BaseLeftNavProps> {
    public static pivotItemsClassName = 'details-view-test-nav-area';
    public render(): JSX.Element {
        const { selectedKey, links } = this.props;

        return (
            <Nav
                className={BaseLeftNav.pivotItemsClassName}
                selectedKey={selectedKey}
                groups={[
                    {
                        links,
                    },
                ]}
                onRenderLink={this.onRenderLink}
                onLinkClick={this.onNavLinkClick}
                styles={{
                    chevronButton: 'hidden',
                }}
            />
        );
    }

    protected onNavLinkClick = (
        event: React.MouseEvent<HTMLElement>,
        item: BaseLeftNavLink,
    ): void => {
        if (item) {
            item.onClickNavLink(event, item);
        }
    };

    protected onRenderLink = (link: BaseLeftNavLink): JSX.Element => {
        return link.onRenderNavLink(link, this.props.renderIcon);
    };
}
