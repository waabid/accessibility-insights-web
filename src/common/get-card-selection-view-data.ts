// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { flatMap, forOwn, keys } from 'lodash';
import {
    CardSelectionStoreData,
    RuleExpandCollapseData,
    RuleExpandCollapseDataDictionary,
} from './types/store-data/card-selection-store-data';

export interface CardSelectionViewData {
    highlightedResultUids: string[]; // page elements to highlight
    selectedResultUids: string[]; // indicates selected cards
    expandedRuleIds: string[];
}

export type GetCardSelectionViewData = (storeData: CardSelectionStoreData) => CardSelectionViewData;

export const getCardSelectionViewData: GetCardSelectionViewData = (storeData: CardSelectionStoreData): CardSelectionViewData => {
    const viewData = getEmptyViewData();

    if (!storeData) {
        return viewData;
    }

    viewData.expandedRuleIds = getRuleIdsOfExpandedRules(storeData.rules);

    if (viewData.expandedRuleIds.length === 0) {
        viewData.highlightedResultUids = getAllResultUids(storeData.rules);
        return viewData;
    }

    viewData.selectedResultUids = getOnlyResultUidsFromSelectedCards(storeData.rules, viewData.expandedRuleIds);

    viewData.highlightedResultUids = viewData.selectedResultUids.length
        ? viewData.selectedResultUids
        : getAllResultUidsFromRuleIdArray(storeData.rules, viewData.expandedRuleIds);

    return viewData;
};

function getEmptyViewData(): CardSelectionViewData {
    const viewData: CardSelectionViewData = {
        highlightedResultUids: [],
        selectedResultUids: [],
        expandedRuleIds: [],
    };

    return viewData;
}

function getRuleIdsOfExpandedRules(ruleDictionary: RuleExpandCollapseDataDictionary): string[] {
    if (!ruleDictionary) {
        return [];
    }

    const expandedRuleIds: string[] = [];

    forOwn(ruleDictionary, (rule, ruleId) => {
        if (rule.isExpanded) {
            expandedRuleIds.push(ruleId);
        }
    });

    return expandedRuleIds;
}

function getAllResultUids(ruleDictionary: RuleExpandCollapseDataDictionary): string[] {
    return getAllResultUidsFromRuleIdArray(ruleDictionary, keys(ruleDictionary));
}

function getAllResultUidsFromRuleIdArray(ruleDictionary: RuleExpandCollapseDataDictionary, ruleIds: string[]): string[] {
    return flatMap(ruleIds, key => getAllResultUidsFromRule(ruleDictionary[key]));
}

function getAllResultUidsFromRule(rule: RuleExpandCollapseData): string[] {
    return keys(rule.cards);
}

function getOnlyResultUidsFromSelectedCards(ruleDictionary: RuleExpandCollapseDataDictionary, ruleIds: string[]): string[] {
    return flatMap(ruleIds, key => getResultUidsFromSelectedCards(ruleDictionary[key]));
}

function getResultUidsFromSelectedCards(rule: RuleExpandCollapseData): string[] {
    const results: string[] = [];

    forOwn(rule.cards, (value, key) => {
        if (value) {
            results.push(key);
        }
    });

    return results;
}
