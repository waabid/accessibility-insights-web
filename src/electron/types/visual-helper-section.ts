// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { ReactFCWithDisplayName } from 'common/react/named-fc';
import { ScreenshotViewProps } from 'electron/views/screenshot/screenshot-view';
import { VirtualKeyboardViewProps } from 'electron/views/virtual-keyboard/virtual-keyboard-view';

export type VisualHelperSectionProps = ScreenshotViewProps & VirtualKeyboardViewProps;

export type VisualHelperSection = ReactFCWithDisplayName<VisualHelperSectionProps>;
