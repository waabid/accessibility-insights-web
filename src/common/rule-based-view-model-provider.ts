// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { HighlightState } from 'common/components/cards/instance-details-footer';
import { CardSelectionViewData } from 'common/get-card-selection-view-data';
import { includes } from 'lodash';
import {
    AllRuleResultStatuses,
    CardResult,
    CardRuleResult,
    CardRuleResultsByStatus,
    CardRuleResultStatus,
    CardsViewModel,
} from './types/store-data/card-view-model';
import { UnifiedResult, UnifiedRule } from './types/store-data/unified-data-interface';

export type GetCardViewData = (
    rules: UnifiedRule[],
    results: UnifiedResult[],
    cardSelectionViewData: CardSelectionViewData,
) => CardsViewModel;

export const getCardViewData: GetCardViewData = (
    rules: UnifiedRule[],
    results: UnifiedResult[],
    cardSelectionViewData: CardSelectionViewData,
): CardsViewModel => {
    if (results == null || rules == null || cardSelectionViewData == null) {
        return null;
    }

    const statusResults = getEmptyStatusResults();
    const ruleIdsWithResultNodes: Set<string> = new Set();

    for (const result of results) {
        const ruleResults = statusResults[result.status];
        const isFailedInstance = result.status === 'fail';
        let ruleResult = getExistingRuleFromResults(result.ruleId, ruleResults);

        if (!ruleResult) {
            const rule = getUnifiedRule(result.ruleId, rules);
            if (!rule) {
                continue;
            }

            const isExpanded = isFailedInstance
                ? includes(cardSelectionViewData.expandedRuleIds, rule.id)
                : false;

            ruleResult = createCardRuleResult(result.status, rule, isExpanded);
            ruleResults.push(ruleResult);
        }

        const isSelected = isFailedInstance
            ? includes(cardSelectionViewData.selectedResultUids, result.uid)
            : false;
        const highlightStatus = cardSelectionViewData.resultHighlightStatus[result.uid];

        ruleResult.nodes.push(createCardResult(result, isSelected, highlightStatus));

        ruleIdsWithResultNodes.add(result.ruleId);
    }

    for (const rule of rules) {
        if (!ruleIdsWithResultNodes.has(rule.id)) {
            statusResults.inapplicable.push(createRuleResultWithoutNodes('inapplicable', rule));
        }
    }

    return {
        cards: statusResults,
        visualHelperEnabled: cardSelectionViewData.visualHelperEnabled,
        allCardsCollapsed: cardSelectionViewData.expandedRuleIds.length === 0,
    };
};

const getExistingRuleFromResults = (
    ruleId: string,
    ruleResults: CardRuleResult[],
): CardRuleResult => {
    const ruleResultIndex: number = getRuleResultIndex(ruleId, ruleResults);

    return ruleResultIndex !== -1 ? ruleResults[ruleResultIndex] : null;
};

const getEmptyStatusResults = (): CardRuleResultsByStatus => {
    const statusResults = {};

    AllRuleResultStatuses.forEach(status => {
        statusResults[status] = [];
    });

    return statusResults as CardRuleResultsByStatus;
};

const getHighlightStatus = (
    isFailedInstance: boolean,
    highlightedResultUids: string[],
    resultUid: string,
) => {
    if (!isFailedInstance) {
        return 'unavailable';
    }

    if (includes(highlightedResultUids, resultUid)) {
        return 'visible';
    }

    return 'hidden';
};

const createCardRuleResult = (
    status: string,
    rule: UnifiedRule,
    isExpanded: boolean,
): CardRuleResult => ({
    id: rule.id,
    status: status,
    nodes: [],
    description: rule.description,
    url: rule.url,
    guidance: rule.guidance,
    isExpanded: isExpanded,
});

const createRuleResultWithoutNodes = (
    status: CardRuleResultStatus,
    rule: UnifiedRule,
): CardRuleResult => ({
    id: rule.id,
    status: status,
    nodes: [],
    description: rule.description,
    url: rule.url,
    guidance: rule.guidance,
    isExpanded: false,
});

const createCardResult = (
    unifiedResult: UnifiedResult,
    isSelected: boolean,
    highlightStatus: HighlightState,
): CardResult => {
    return {
        ...unifiedResult,
        isSelected,
        highlightStatus,
    };
};

const getUnifiedRule = (id: string, rules: UnifiedRule[]): UnifiedRule =>
    rules.find(rule => rule.id === id);

const getRuleResultIndex = (ruleId: string, ruleResults: CardRuleResult[]): number =>
    ruleResults.findIndex(result => result.id === ruleId);
