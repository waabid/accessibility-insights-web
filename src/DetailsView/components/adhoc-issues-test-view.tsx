// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedFC } from '../../common/react/named-fc';
import { DetailsListIssuesView, DetailsListIssuesViewDeps, DetailsListIssuesViewProps } from './details-list-issues-view';
import { TargetPageChangedView } from './target-page-changed-view';

export type AdhocIssuesTestViewDeps = DetailsListIssuesViewDeps;

export type AdhocIssuesTestViewProps = DetailsListIssuesViewProps;

export const AdhocIssuesTestView = NamedFC<AdhocIssuesTestViewProps>('AdhocIssuesTestView', ({ children, ...props }) => {
    if (props.tabStoreData.isChanged) {
        return createTargetPageChangedView(props);
    }

    return <DetailsListIssuesView {...props} />;
});

function createTargetPageChangedView(props: AdhocIssuesTestViewProps): JSX.Element {
    const selectedTest = props.selectedTest;
    const scanData = props.configuration.getStoreData(props.visualizationStoreData.tests);
    const clickHandler = props.clickHandlerFactory.createClickHandler(selectedTest, !scanData.enabled);

    return (
        <TargetPageChangedView
            displayableData={props.configuration.displayableData}
            visualizationType={selectedTest}
            toggleClickHandler={clickHandler}
            featureFlagStoreData={props.featureFlagStoreData}
        />
    );
}
