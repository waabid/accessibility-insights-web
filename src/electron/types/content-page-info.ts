// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ReactFCWithDisplayName } from 'common/react/named-fc';
import { ResultsFilter } from 'common/types/results-filter';
import { TestingContentProps } from 'electron/platform/android/testing-content';
import { VisualHelperSection } from 'electron/types/visual-helper-section';
import { ReflowCommandBarProps } from 'electron/views/results/components/reflow-command-bar';
import { LeftNavItemKey } from './left-nav-item-key';

export type StartOverButtonSettings = {
    onClick: () => void;
    disabled: boolean;
};

export type ContentPageInfo = {
    title: string;
    allowsExportReport: boolean;
    description?: JSX.Element;
    instancesSectionComponent: ReactFCWithDisplayName<TestingContentProps>;
    resultsFilter: ResultsFilter;
    visualHelperSection: VisualHelperSection;
    startOverButtonSettings: (props: ReflowCommandBarProps) => StartOverButtonSettings;
};

export type ContentPagesInfo = {
    [key in keyof LeftNavItemKey]: ContentPageInfo;
};
