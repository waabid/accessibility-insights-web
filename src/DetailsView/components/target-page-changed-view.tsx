// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FeatureFlags } from 'common/feature-flags';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { Toggle } from 'office-ui-fabric-react/lib/Toggle';
import * as React from 'react';

import { DisplayableVisualizationTypeData } from '../../common/configs/visualization-configuration-factory';
import { NamedFC } from '../../common/react/named-fc';
import { VisualizationType } from '../../common/types/visualization-type';

export interface TargetPageChangedViewProps {
    visualizationType: VisualizationType;
    displayableData: DisplayableVisualizationTypeData;
    toggleClickHandler: (event) => void;
    featureFlagStoreData: FeatureFlagStoreData;
}

export const TargetPageChangedView = NamedFC<TargetPageChangedViewProps>('TargetPageChangedView', props => {
    const { title = '', toggleLabel = '', subtitle } = props.displayableData;

    const oldText = 'The target page was changed. Use the toggle to enable the visualization in the current target page.';
    const cardsUIText = 'The target page has changed. Use the start over button to scan this new target page.';
    const displayedText = props.featureFlagStoreData[FeatureFlags.universalCardsUI] ? cardsUIText : oldText;

    return (
        <div className="target-page-changed">
            <h1>{title}</h1>
            <div className="target-page-changed-subtitle">{subtitle}</div>
            <Toggle
                onText="On"
                offText="Off"
                checked={false}
                onClick={props.toggleClickHandler}
                label={toggleLabel}
                className="details-view-toggle"
            />
            <p>{displayedText}</p>
        </div>
    );
});
