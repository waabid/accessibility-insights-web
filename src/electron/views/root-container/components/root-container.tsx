// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import {
    withStoreSubscription,
    WithStoreSubscriptionDeps,
} from 'common/components/with-store-subscription';
import { NamedFC } from 'common/react/named-fc';
import { CardSelectionStoreData } from 'common/types/store-data/card-selection-store-data';
import { DetailsViewStoreData } from 'common/types/store-data/details-view-store-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { UnifiedScanResultStoreData } from 'common/types/store-data/unified-data-interface';
import { UserConfigurationStoreData } from 'common/types/store-data/user-configuration-store';
import {
    NarrowModeDetector,
    NarrowModeDetectorDeps,
} from 'DetailsView/components/narrow-mode-detector';
import { AndroidSetupStoreData } from 'electron/flux/types/android-setup-store-data';
import { LeftNavStoreData } from 'electron/flux/types/left-nav-store-data';
import { ScanStoreData } from 'electron/flux/types/scan-store-data';
import { WindowStateStoreData } from 'electron/flux/types/window-state-store-data';
import {
    MainApplicationView,
    MainApplicationViewDeps,
    MainApplicationViewProps,
} from 'electron/views/automated-checks/main-application-view';
import {
    DeviceConnectViewContainer,
    DeviceConnectViewContainerDeps,
} from 'electron/views/device-connect-view/components/device-connect-view-container';
import * as React from 'react';
import './root-container.scss';

export type RootContainerDeps = WithStoreSubscriptionDeps<RootContainerState> &
    DeviceConnectViewContainerDeps &
    MainApplicationViewDeps &
    NarrowModeDetectorDeps;

export type RootContainerProps = {
    deps: RootContainerDeps;
    storeState: RootContainerState;
};

export type RootContainerState = {
    windowStateStoreData: WindowStateStoreData;
    userConfigurationStoreData: UserConfigurationStoreData;
    scanStoreData: ScanStoreData;
    unifiedScanResultStoreData: UnifiedScanResultStoreData;
    cardSelectionStoreData: CardSelectionStoreData;
    detailsViewStoreData: DetailsViewStoreData;
    featureFlagStoreData: FeatureFlagStoreData;
    androidSetupStoreData: AndroidSetupStoreData;
    leftNavStoreData: LeftNavStoreData;
};

export const RootContainerInternal = NamedFC<RootContainerProps>('RootContainerInternal', props => {
    const { storeState, ...rest } = props;

    const childProps: Omit<MainApplicationViewProps, 'narrowModeStatus'> = {
        ...storeState,
        ...rest,
    };

    if (storeState.windowStateStoreData.routeId === 'resultsView') {
        return (
            <NarrowModeDetector
                deps={props.deps}
                isNarrowModeEnabled={true}
                Component={MainApplicationView}
                childrenProps={childProps}
            />
        );
    }

    return <DeviceConnectViewContainer {...storeState} {...rest} />;
});

export const RootContainer = withStoreSubscription<RootContainerProps, RootContainerState>(
    RootContainerInternal,
);
