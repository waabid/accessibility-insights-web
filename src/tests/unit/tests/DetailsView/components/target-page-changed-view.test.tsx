// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FeatureFlags } from 'common/feature-flags';
import { shallow } from 'enzyme';
import * as React from 'react';

import { DisplayableVisualizationTypeData } from '../../../../../common/configs/visualization-configuration-factory';
import { VisualizationType } from '../../../../../common/types/visualization-type';
import { TargetPageChangedView, TargetPageChangedViewProps } from '../../../../../DetailsView/components/target-page-changed-view';

describe('TargetPageChangedView', () => {
    it('renders without optional subtitle', () => {
        testRenderWithOptionalSubtitle(undefined, false);
    });

    it('renders with feature flag enabled', () => {
        testRenderWithOptionalSubtitle(undefined, true);
    });

    it('renders with optional subtitle', () => {
        testRenderWithOptionalSubtitle(<span>test subtitle content</span>, false);
    });

    function testRenderWithOptionalSubtitle(subtitle: JSX.Element, featureFlagEnabled: boolean): void {
        const visualizationType = VisualizationType.Landmarks;
        const clickHandlerStub: () => void = () => {};
        const displayableData = {
            title: 'test title',
            toggleLabel: 'test toggle label',
            subtitle,
        } as DisplayableVisualizationTypeData;

        const props: TargetPageChangedViewProps = {
            visualizationType,
            displayableData,
            toggleClickHandler: clickHandlerStub,
            featureFlagStoreData: {
                [FeatureFlags.universalCardsUI]: featureFlagEnabled,
            },
        };

        const wrapped = shallow(<TargetPageChangedView {...props} />);

        expect(wrapped.getElement()).toMatchSnapshot();
    }
});
