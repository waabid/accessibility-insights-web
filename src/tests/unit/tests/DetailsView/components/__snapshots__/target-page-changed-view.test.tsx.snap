// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TargetPageChangedView renders with feature flag enabled 1`] = `
<div
  className="target-page-changed"
>
  <h1>
    test title
  </h1>
  <div
    className="target-page-changed-subtitle"
  />
  <StyledToggleBase
    checked={false}
    className="details-view-toggle"
    label="test toggle label"
    offText="Off"
    onClick={[Function]}
    onText="On"
  />
  <p>
    The target page has changed. Use the start over button to scan this new target page.
  </p>
</div>
`;

exports[`TargetPageChangedView renders with optional subtitle 1`] = `
<div
  className="target-page-changed"
>
  <h1>
    test title
  </h1>
  <div
    className="target-page-changed-subtitle"
  >
    <span>
      test subtitle content
    </span>
  </div>
  <StyledToggleBase
    checked={false}
    className="details-view-toggle"
    label="test toggle label"
    offText="Off"
    onClick={[Function]}
    onText="On"
  />
  <p>
    The target page was changed. Use the toggle to enable the visualization in the current target page.
  </p>
</div>
`;

exports[`TargetPageChangedView renders without optional subtitle 1`] = `
<div
  className="target-page-changed"
>
  <h1>
    test title
  </h1>
  <div
    className="target-page-changed-subtitle"
  />
  <StyledToggleBase
    checked={false}
    className="details-view-toggle"
    label="test toggle label"
    offText="Off"
    onClick={[Function]}
    onText="On"
  />
  <p>
    The target page was changed. Use the toggle to enable the visualization in the current target page.
  </p>
</div>
`;
